public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -754,6 +778,25 @@ export class LokiDatasource
       return addLabelToQuery(queryExpr, key, value, operator, true);
     }
   }
+
+  // Used when running queries through backend
+  filterQuery(query: LokiQuery): boolean {
+    if (query.hide || query.expr === '') {
+      return false;
+    }
+    return true;
+  }
+
+  // Used when running queries through backend
+  applyTemplateVariables(target: LokiQuery, scopedVars: ScopedVars): Record<string, any> {
+    // We want to interpolate these variables on backend
+    const { __interval, __interval_ms, ...rest } = scopedVars;
+
+    return {
+      ...target,
+      expr: this.templateSrv.replace(target.expr, rest, this.interpolateQueryExpr),
```

Comment:
Do we need to interpolate the `volumeQuery` or `query` as well?

Reply:
- `volumeQuery` we don't need to interpolate. that's a `boolean` that is not user-modifiable
- `query` ... i don't know if this attribute is ever used. i checked how the query is currently created in the frontend in https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/datasource.ts#L238 and it seems it only uses `.expr` and `.resolution` (`.resolution` cannot contain a variable)

i added `legendFormat` interpolation, hopefully this should cover all of them. thanks for looking into this part!